[Android] Exec wake-up routine only after hdmi handshake

hugbug opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
When device wakes up after sleep the wake-up jobs are executed. They include reinitialisation of audio engine. To avoid errors in audio engine we wait for hdmi handshake to complete before executing wake-up jobs.
This PR improves my previous PRs #24369 and #24484.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
PR #24369 has fixed an issue with audio passthrough not working after sleep/wakeup. The issue has affected Fire TV and Shield devices, especially when used with AVRs. The fix worked well for me until I had to change passthrough and HDMI-CEC settings on my AVR. The change was necessary to fix an issue with AVR waking up by TV when it shouldn't.

After the changes of HDMI-settings on AVR the problem with audio passthrough on Fire TV has returned. The fix of PR #24369 didn't work anymore. The Kodi debug logs indicated an error when resuming audio engine during wakeup jobs (ACTION_SCREEN_ON):

>2024-02-03 12:01:52.852 T:5647    debug <general>: CXBMCApp::onReceive - Got intent. Action: android.intent.action.SCREEN_ON
>2024-02-03 12:01:52.852 T:5647     info <general>: Got device wakeup intent
>2024-02-03 12:01:53.268 T:5674     info <general>: OnWake: Running resume jobs
>...
>2024-02-03 12:01:58.269 T:5674    error <general>: ActiveAE::Resume - failed to init

Later, when ACTION_HDMI_AUDIO_PLUG event is received, an attempt is made to reinitialise audio engine but it fails:
>2024-02-03 12:01:58.463 T:5647    debug <general>: CXBMCApp::onReceive - Got intent. Action: android.media.action.HDMI_AUDIO_PLUG
>2024-02-03 12:01:58.463 T:5647    debug <general>: -- HDMI is plugged in: yes
>2024-02-03 12:01:58.463 T:5647    debug <general>: CWinSystemAndroid::SetHdmiState: state: 1
>2024-02-03 12:01:58.463 T:5647    debug <general>: CXBMCApp::onReceive: Reset audio engine
>...
>2024-02-03 12:01:58.500 T:5691  warning <general>: CActiveAE::StateMachine - signal: 3 from port: OutputControlPort not handled for state: 2

Full log containing error: https://paste.kodi.tv/opurigawed.kodi

### Possible approaches to fix the issue

1. The failed resuming of audio engine sets it into an error state in which it doesn't process device-change events. I guess it would be possible to extend/rework audio engine to be able to handle device change events even in error state. That requires a rework of classes used globally on all platforms.

2. Audio engine is resumed in wakeup-jobs. The jobs are executed in power management module when event ACTION_SCREEN_ON is processed in Android app. We could change the wakeup-jobs management and not resume audio engine there. Instead we could resume it separately later, when ACTION_HDMI_AUDIO_PLUG  is received by our Android app. This code change requires reworking of general power management classes and would also affect other platforms.

3. The less intrusive fix is to run all wakeup-jobs on ACTION_HDMI_AUDIO_PLUG event instead of ACTION_SCREEN_ON. That requires changes only in Android app class. That's the solution proposed in this PR.

### Implementation details
Event ACTION_HDMI_AUDIO_PLUG is processed only on certain device types. If we can use the event we run wakeup-jobs in this event, otherwise in ACTION_SCREEN_ON (old behaviour). In order to execute the required code in two different events I extracted the code from ACTION_SCREEN_ON into a new routine `CXBMCApp::OnWakeup()` (no actual changes to the code there).

I also moved the code from ACTION_SCREEN_OFF event to a new routine `CXBMCApp::OnSleep()`. That was technically not necessary but makes the code cleaner and keeps the both related code parts nearby.

New class member `m_aeReset` was introduced to keep the fix #24484 working.

I've been using the fix for the last 10 days without issues on my Fire TV Stick 4K Max 2nd gen.

Kodi-log after fix shows no error during resuming of audio engine: https://paste.kodi.tv/racuzilaju.kodi

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Fire TV Stick 4K Max 2nd gen, connected to AVR Denon AVC-X4800H, connected to TV LG C1. Fire TV is configured to stop outputting signal on sleep. Fire TV and Kodi are configured to output all audio formats as passthrough, using Kodi IEC packer. HDMI-passthrough and HDMI-Control (CEC) are disabled on AVR.

1. Start Kodi.
2. Make sure audio passthrough is configured in Kodi settings, using Kodi IEC packer.
3. Playback a video with multichannel audio such as Dolby Digital or DTS, check on AVR that the incoming signal is correct.
4. Stop video playback.
5. Switch off all devices using on/off button on Fire TV remote. If Fire TV is properly configured the AVR and TV will both switch off.
6. Wait a few seconds for sleep mode to activate.
7. Press on/off button on Fire TV remote to wake up Fire TV and switch on AVR and TV.
8. Kodi should be active.
9. Try to playback the same video.
10. Check incoming signal on AVR.
11. Go to Audio settings in Kodi. Check that Kodi IEC packer is still selected as audio passthrough device.

### Test results

#### Before fix

On step 10 AVR shows input signal as "PCM Stereo". On step 11 "Kodi IEC packer" disappears from the list of audio devices. Android IEC packer is shown instead (it doesn't work properly on Fire TV).

#### After fix

On step 10 AVR shows the same input signal as on step 3. On step 11 "Kodi IEC packer" is in the list of audio devices and it is selected as output device. Audio passthrough works correctly.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
More robust handling of sleep/wakeup states on Android fixes some remaining audio passthrough issues, happened with certain device combinations, in particular Fire TV connected to AVR.

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

hugbug committed:
[Android] Exec wake-up routine only after hdmi handshake

When device wakes up after sleep the wake-up jobs are executed. They include reinitialisation of audio engine. To avoid errors in audio engine we wait for hdmi handshake to complete before executing wake-up jobs.

thexai commented on xbmc/platform/android/activity/XBMCApp.cpp:
```diff
@@ -1433,6 +1425,29 @@ void CXBMCApp::onReceive(CJNIIntent intent)
 
+void CXBMCApp::OnSleep()
+{
+  CLog::Log(LOGINFO, "CXBMCApp::OnSleep");
```
```suggestion
  CLog::Log(LOGDEBUG, "CXBMCApp::OnSleep");
```
Another similar log line already exists: "Got device sleep intent"

thexai commented on xbmc/platform/android/activity/XBMCApp.cpp:
```diff
@@ -1433,6 +1425,29 @@ void CXBMCApp::onReceive(CJNIIntent intent)
+
+void CXBMCApp::OnWakeup()
+{
+  CLog::Log(LOGINFO, "CXBMCApp::OnWakeup");
```
```suggestion
  CLog::Log(LOGDEBUG, "CXBMCApp::OnWakeup");
```

Same: is a bit redundant of "Got device wakeup intent", keep in debug only

fritsch commented:
General thoughts:
One major issue with the hdmi hotplug event is, that it simply does not come on e.g. "old FireTV" version, which was also the reason for a "maximum" timer for the Adjust Refreshrate Code.

Also entirely not sure: How does this affect Smartphones?

In general we cannot assume that we get the hdmi hotplug in general.

hugbug committed:
[Android] Lowered logging level

Another similar log lines already exist: "Got device sleep intent" and "Got device wakeup intent", keep in debug only.

Co-Authored-By: thexai <58434170+thexai@users.noreply.github.com>

hugbug committed:
[Android] Check if ACTION_HDMI_AUDIO_PLUG works

The wakeup-routine now relies on ACTION_HDMI_AUDIO_PLUG only when this intent is really supported by device.

hugbug commented:
Thank you for review and suggestions.

On start Kodi registers for ACTION_HDMI_AUDIO_PLUG intent:
https://github.com/xbmc/xbmc/blob/32f33872329ae4afecdd16769ddad2e1634319d8/xbmc/platform/android/activity/XBMCApp.cpp#L246-L248

Depending on `m_hdmiSource` the wakeup-jobs are executed either in ACTION_HDMI_AUDIO_PLUG or in ACTION_SCREEN_ON. I thought that if we register for ACTION_HDMI_AUDIO_PLUG it will fire up, but I understand the concerns that this might not always be the case.

## Improvement

During testing I noticed that one event ACTION_HDMI_AUDIO_PLUG is received immediately after we register for it (you can't even see it in kodi.log, only in logcat). We don't process it at such early app loading stage but I think we can use that as an indication that the event is supported by the system. I've made a change to do exactly this.

Here we set the flag
https://github.com/xbmc/xbmc/blob/32f33872329ae4afecdd16769ddad2e1634319d8/xbmc/platform/android/activity/XBMCApp.cpp#L1334-L1336

And then use it:
https://github.com/xbmc/xbmc/blob/32f33872329ae4afecdd16769ddad2e1634319d8/xbmc/platform/android/activity/XBMCApp.cpp#L1357-L1370

If on some system event ACTION_HDMI_AUDIO_PLUG isn't sent immediately as we expect, nothing bad happens. The first wakeup-routine will be executed on ACTION_SCREEN_ON, later wakeups will be executed in ACTION_HDMI_AUDIO_PLUG.

I hope this covers all possible cases including old FireTV and smartphones.

Tested this new version and it works fine with my device. As a second test I commented out the line  `m_supportsHdmiAudioPlug = true;` to simulate not receiving of the event. The bug returned (as expected) but everything else worked fine as it should.

fritsch commented:
Thanks very much - I think there is no other way around besides testing it.

thexai reviewed (approved):
Tested on Shield latest version.

Tested various use cases: power off from Shield remote (powers off TV and AVR) and power off from TV remote (also powers off AVR) and in all cases audio device "Kodi IEC packer" is present after wake-up... although in my setup I have never had these problems previously.

Change seems robust to handle all cases.

a1rwulf reviewed (approved):
Looks good to me.
I don't have time to do tests atm, but I have seen this issue in real life.
Shield is also my daily driver, so I rely on @thexai 's tests.

a1rwulf commented:
I think I've seen it in a slightly different scenario though, which I mitigated by adding a desktop resolution for Android in order to skip a potential modeset on app startup and just reuse whatever is set in the Android display settings.

fritsch reviewed (approved):
Thanks again.
